spec/kintsugi_spec.rb

```diff
@@ -0,0 +1,98 @@
+# Copyright (c) 2021 Lightricks. All rights reserved.
+# Created by Ben Yohay.
+# frozen_string_literal: true
+
+require "git"
+require "json"
+require "rspec"
+require "tempfile"
+require "tmpdir"
+
+require "kintsugi"
+
+shared_examples "tests" do |project_name|
+  let(:temporary_directories_paths) { [] }
+
+  after do
+    temporary_directories_paths.each do |directory_path|
+      FileUtils.remove_entry(directory_path)
+    end
+  end
+
+  it "resolves conflicts" do
+    git_directory_path = make_temp_directory
+
+    git = Git.init(git_directory_path)
+    File.write(File.join(git_directory_path, ".gitattributes"), "*.pbxproj merge=Unset")
+
+    project_path = File.join(git_directory_path, project_name)
+    project = Xcodeproj::Project.new(project_path)
+    project.save
+
+    git.add(project_path)
+    git.commit("Initial project")
+
+    project.new_target("com.apple.product-type.library.static", "foo", :ios)
+    project.save
+
+    git.add(:all => true)
+    git.commit("Add target foo")
+    first_commit_hash = git.revparse("HEAD")
+
+    git.checkout("HEAD^")
+    project = Xcodeproj::Project.open(project_path)
+    project.new_target("com.apple.product-type.library.static", "bar", :ios)
+    project.save
+    git.add(:all => true)
+    git.commit("Add target bar")
+
+    `git -C #{git_directory_path} rebase #{first_commit_hash} &> /dev/null`
+    Kintsugi::run([File.join(project.path, "project.pbxproj")])
+
+    project = Xcodeproj::Project.open(project_path)
+    expect(project.targets.count).to equal(2)
+  end
+
+  it "performs three way merge" do
+    directory_path = make_temp_directory
+
+    base_project = create_new_project_at_path(File.join(directory_path, "base.xcodeproj"))
+    ours_project = create_new_project_at_path(File.join(directory_path, "ours.xcodeproj"))
+    theirs_project = create_new_project_at_path(File.join(directory_path, "theirs.xcodeproj"))
+    original_project = create_new_project_at_path(File.join(directory_path, project_name))
+
+    ours_project.new_target("com.apple.product-type.library.static", "foo", :ios)
+    ours_project.save
+
+    theirs_project.new_target("com.apple.product-type.library.static", "bar", :ios)
+    theirs_project.save
+
+    Kintsugi::run(["driver",
+                   base_project.path + "project.pbxproj",
+                   ours_project.path + "project.pbxproj",
+                   theirs_project.path + "project.pbxproj",
+                   original_project.path + "project.pbxproj"])
+
+    # Changes are supposed to be in `ours_project` after driver is called.
+    ours_project = Xcodeproj::Project.open(ours_project.path)
+    expect(Xcodeproj::Project.open(ours_project.path).targets.count).to equal(2)
+  end
+
```

Comment:
add tests for errors - make sure the state of the folder in the end is what it should be.
If you can, also add tests for the driver

Reply:
Added a test that checks that after Kintsugi failed to resolve a conflict the project wasn't touched and is still has conflicts.